packages/analytics-browser/src/plugins/session-handler.ts

```diff
@@ -0,0 +1,55 @@
+import { BeforePlugin, BrowserClient, BrowserConfig, Event, PluginType } from '@amplitude/analytics-types';
+
+export const START_SESSION_EVENT = 'session_start';
+export const END_SESSION_EVENT = 'session_end';
+
+export const sessionHandlerPlugin = (): BeforePlugin => {
+  // browserConfig is defined in setup() which will always be called first
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-ignore
+  let browserConfig: BrowserConfig;
+  // amplitude is defined in setup() which will always be called first
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-ignore
+  let amplitude: BrowserClient;
+
+  const name = '@amplitude/plugin-session-handler';
+
+  const type = PluginType.BEFORE;
+
+  const setup = async (config: BrowserConfig, client: BrowserClient) => {
+    browserConfig = config;
+    amplitude = client;
+  };
+
+  const execute = async (event: Event) => {
+    const now = Date.now();
+
+    if (event.event_type === START_SESSION_EVENT || event.event_type === END_SESSION_EVENT) {
+      browserConfig.lastEventTime = now;
+      return event;
+    }
+
+    const lastEventTime = browserConfig.lastEventTime || now;
+    const timeSinceLastEvent = now - lastEventTime;
+
+    if (timeSinceLastEvent > browserConfig.sessionTimeout) {
+      // assigns new session
+      amplitude.setSessionId(now);
```

Comment:
Curious about the order of the events, would this show before start session event in this case?

Reply:
The order of events in payload request are definitely not in the right order, but the session events themselves have the time property manipulated so analytics would see them in the intended order. I added some tests on that here: https://github.com/amplitude/Amplitude-TypeScript/blob/5a1f0e12fcd5d105d8764e7c2230cdc49ac1f6ef/packages/analytics-browser-test/test/index.test.ts#L926-L929

Reply:
I see, that sounds good to me.